appstub.mod/debugger_mt.stdio.bmx

```diff
@@ -409,95 +411,119 @@ Function DebugScopeKind$( scope:Int Ptr )
-?Not ptr64
-Function DebugDerefPointer$(decl:Int Ptr, pointer:Int Ptr)
-? 
-	Local tipe$=DebugDeclType( decl )
-	Local datatype$ = tipe[..tipe.Find(" ",0)]
-	Local start:Int
-	Local count:Int
+Function DebugDerefPointer:String(decl:Int Ptr, pointer:Byte Ptr)
+	Const derefFailure:String = "{?}"
+	Const derefSymbol:String = "->"
 	
-	Repeat
-		start = tipe.Find("Ptr",start+1)
-		If start = -1 Exit
-		count :+ 1
-	Forever
-
-	For Local i:Int = 0 Until count
-		' Null
-		If pointer = 0 Then
-			Return " {-}"
+	Local declType:String = DebugDeclType(decl)
+	Local dataType:String = declType
+	Local ptrDepth:Int = 0
+	While dataType.EndsWith(" Ptr")
+		dataType = dataType[..dataType.length - " Ptr".length]
+		ptrDepth :+ 1
+	Wend
+	
+	Local result:String = ""
+	
+	
+	? Win32
+	Extern "Win32"
+		Function GetCurrentProcess:Byte Ptr() = "HANDLE GetCurrentProcess(void)!"
+		Function ReadProcessMemory:Int(hProcess:Byte Ptr, lpBaseAddress:Byte Ptr, lpBuffer:Byte Ptr, nSize:Size_T, lpNumberOfBytesRead:Size_T Ptr) = "BOOL ReadProcessMemory(HANDLE, LPCVOID, LPVOID, SIZE_T, SIZE_T*)!"
+	End Extern
+	
+	Local dataSize:Size_T
+	Select dataType
+ 		Case "Byte"      dataSize = SizeOf(Byte Null)
+		Case "Short"     dataSize = SizeOf(Short Null)
+		Case "Int"       dataSize = SizeOf(Int Null)
+		Case "UInt"      dataSize = SizeOf(UInt Null)
+		Case "Long"      dataSize = SizeOf(Long Null)
+		Case "ULong"     dataSize = SizeOf(ULong Null)
+		Case "Size_T"    dataSize = SizeOf(Size_T Null)
+		Case "Float"     dataSize = SizeOf(Float Null)
+		Case "Double"    dataSize = SizeOf(Double Null)
+	? Win32 And Ptr64
+		Case "Float64"   dataSize = SizeOf(Float64 Null)
+		Case "Float128"  dataSize = SizeOf(Float128 Null)
+		Case "Double128" dataSize = SizeOf(Double128 Null)
+		Case "Int128"    dataSize = SizeOf(Int128 Null)
+	? Win32	
+		Case "WParam"    dataSize = SizeOf(WParam Null)
+		Case "LParam"    dataSize = SizeOf(LParam Null)
+		Default          dataSize = 0 ' cannot dereference this
+	EndSelect
+	
+	Local processHandle:Byte Ptr = GetCurrentProcess()
+	Local buffer:Byte Ptr = MemAlloc(Max(dataSize, SizeOf(Byte Ptr Null)))
+	(Byte Ptr Ptr buffer)[0] = Null
+	
+	pointer = (Byte Ptr Ptr pointer)[0]
+	For Local i:Int = 1 To ptrDepth - 1
+		Local success:Int = ReadProcessMemory(processHandle, pointer, buffer, Size_T SizeOf(Byte Ptr Null), Null)
+		If Not success Then
+			MemFree buffer
+			result :+ derefSymbol + derefFailure
+			Return result
 		End If
-?ptr64
-		pointer = Long Ptr (Varptr pointer)[0]
-?Not ptr64
-		pointer = Int Ptr (Varptr pointer)[0]
-?
+		pointer = (Byte Ptr Ptr buffer)[0]
+	? Win32 And Not Ptr64
+		result :+ derefSymbol + "$" + ToHex(Int pointer)
+	? Win32 And Ptr64
+		result :+ derefSymbol + "$" + ToHex(Long pointer)
+	? Win32
 	Next
-
-	' make sure the final pointer is not null
-	If pointer = 0 
-		Return " {-}"
-	EndIf
-
+	
+	Local success:Int
+	If dataSize > 0 Then
+		success = ReadProcessMemory:Int(processHandle, pointer, buffer, dataSize, Null)

```

Comment:
Wait, I didn't even notice this until now - that's actually valid syntax? :open_mouth: